pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,367 @@
+				err = bfs.fetchData()
+			case ModeOidSearch:
+				err = bfs.fetchDataWithOid()
+			default:
+				bfs.log.Error("Invalid mode specified", zap.String("mode", bfs.cfg.Mode))
+				close(bfs.quit)
+				return
+			}
+
+			if err != nil {
+				bfs.log.Error("Fetch operation failed", zap.Error(err))
+				close(bfs.quit)
+				return
+			}
+		}()
+	}
+}
+
+// Shutdown implements the Service interface.
+func (bfs *Service) Shutdown() {
+	if bfs.started.CompareAndSwap(true, false) {
+		bfs.log.Info("Shutting down Block Fetcher Service")
+		close(bfs.quit)
+		<-bfs.toExitCh
+	}
+}
+
+// IsActive implements the Service interface.
+func (bfs *Service) IsActive() bool {
+	return bfs.started.Load()
+}
+
+func (bfs *Service) fetchData() error {
+	startIndex := bfs.chain.BlockHeight()
+	batchSize := uint32(bfs.cfg.BatchSize)
+
+	for {
+		select {
+		case <-bfs.quit:
+			bfs.log.Info("Stopping fetchData operation.")
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(bfs.cfg.BlockAttribute, fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter(bfs.cfg.BlockAttribute, fmt.Sprintf("%d", startIndex+batchSize-1), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			blockOids, err := bfs.search(prm)
+			bfs.log.Info(fmt.Sprintf("Found %d blocks from index %d to %d", len(blockOids), startIndex, startIndex+batchSize-1))
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to find block with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(blockOids) == 0 {
+				bfs.log.Info(fmt.Sprintf("No block found with index %d, stopping.", startIndex))
+				return errors.New("no block found")
+			}
+			err = bfs.fetchAndProcessBlocks(blockOids)
+			if err != nil {
+				return err
+			}
+			startIndex += batchSize
+		}
+	}
+}
+
+func (bfs *Service) fetchAndProcessBlocks(blockOids []oid.ID) error {
+	var wg sync.WaitGroup
+
+	for _, oidBlock := range blockOids {
+		wg.Add(1)
+		go func(oidBlock oid.ID) {
+			defer wg.Done()
+
+			rc, err := bfs.get(oidBlock.String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to fetch block %s", oidBlock.String()), zap.Error(err))
+				return
+			}
+
+			bfs.log.Info(fmt.Sprintf("Fetched block with OID %s", oidBlock.String()))
+
+			err = bfs.processBlock(rc)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to process block with OID %s", oidBlock.String()), zap.Error(err))
+			}
+		}(oidBlock)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (bfs *Service) fetchDataWithOid() error {
+	startIndex := bfs.chain.BlockHeight() / 2000
+	skip := bfs.chain.BlockHeight() % 2000
+
+	for {
+		select {
+		case <-bfs.quit:
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(bfs.cfg.OidAttribute, fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter(bfs.cfg.OidAttribute, fmt.Sprintf("%d", startIndex), object.MatchNumLE)
```

Comment:
OK, here you may use the exact equality filter, because you need an exact batch of OIDs.